add doc on performance and scalability (#342)
openservicemesh/osm-docs: content/docs/guides/ha_scale/perf-scale.md

## Testing environment

This test is performed on a Kubernetes cluster provisioned and managed by [Microsoft Azure Kubernetes Service](https://azure.microsoft.com/en-us/services/kubernetes-service/). The cluster has 100 Standard_DS2_v2 nodes (each node has 2 vCPUs and 7GB memory). It is ensured that there is only one OSM controller pod running with default resource quota during the test. OSM is configured in non-permissive traffic mode.

## Mesh topology

The testing mesh contains two types of service. One is a load generator (modified from [wrk2](https://github.com/giltene/wrk2)), which can send requests at specified rate, load balanced among all other services in the mesh. The load generator serves as the client side of a request. Throughout our test, there will be at most one load generator service. The other type of service is a simple echo application that replies with whatever it receives from an inbound request. It serves as the server side of a request. We can deploy multiple copies of the echo application in the mesh. In short, the mesh topology is a typical 1-to-N service architecture.

<p align="center">
  <img src="/docs/images/perf/mesh-topo.png" width="650"/>
</p>

> Note that since both types of service are in the mesh and permissive mode is disabled, thus each request will be authenticated via mTLS. It means that requests go through the Envoy sidecar in both the client and the server.

Reviewer: ```suggestion
Note that the sidecar is present both on the client and server, which intercepts all traffic and applies mTLS
```

## Metrics

> The test result includes a lot of metrics collected. A clear definition of each metric can help avoid ambiguity. Below are the metric definitions we will use throughout this documentation:

Reviewer: ```suggestion
Below are the metric definitions we collect and will use throughout this documentation:
```

* Control plane

>     * Peak time CPU/Memory: the maximum CPU/memory usage during the testing process.

Reviewer: can we reword this to just call it max usage? then we don't even need a definition.

>     * Stable time CPU/Memory usage: the average CPU/memory usage when the control plane is off the peak time and becomes stable.

Reviewer: same for this to call it average. let's also change the graph labels

Reviewer: Providing more context on why, calling it "peak time" is a bit misleading, since it implies that there is a time in which there is a peak, and a time in which there is a max. When in reality, there is not a specific time, and both can happen simultaneously

Author: Definition for control plane seems obvious. But some metrics in data plane might have different understanding. E.g. Max usage can be interpreted as the avg of max usage of all sidecars, or the max usage of all sidecars. I suggest we don't completely remove this section.

Author: Regarding the "peak time", in our test workflow, we define the mesh scale. The peak time usage is a good indication of the worst case that the mesh will encounter. In the real case, if we slowly adds workload to the mesh and reach a certain scale, there is unlikely to have a spike. But this result does not mean much for evaluating OSM.

* Data plane sidecar

>     * Peak time CPU/Memory: the maximum CPU/memory usage among all sidecar containers during the testing process.

Reviewer: Following up on the below comment, Instead of "during the testing process" we should be explicit, ie: "during the first 60s while pods are initializing".

    * Stable time CPU/Memory usage: the average CPU/memory usage of all sidecar containers after they become stable.
* Request latency: request roundtrip time. P50, P90, P99 and P999 of request latency are measured.

> The reason that peak time and stable time are measured individually is because of how the services are deployed. All the echo services will be deployed and added to the mesh at the same time. Therefore, the control plane receives a load spike after the deployment starts. All the Envoy sidecars are connecting to the control plane simutaneously. After that, the control plane becomes much more idle, since there is no more change to the mesh topology.

Reviewer: Ok so this makes sense, but I think the above is misleading. If we want to delineate metrics during periods of high churn, lets call that out in the definition, and not as a note here :)

The data plane should largely be unaffected by this churn, so let's just drop peak/stable terminology for that altogether.

The control plane we an call that out, but I don't think it's useful to plot them on the same graph. Let's separate that out is a completely different scenario (there may be differing opinions on this, let's get others' feedback)

Reviewer: Adding more context to the graph separation:

I would think of that as average and max (which is a valid thing to do **if** there are multiple backends we are measuring (true for sidecar, not currently true for osm-controller, although a future scenario may change that). So that's what I expect when I see that graph, and why I think we should separate it.

## Performance

> In this test, we focus on the request overhead added by OSM as well as the computation resource consumptions. The RPS level of the load generator is calculated as the desired per pod RPS multiplies the number of server side pods. We test the per pod RPS at 500, 1000, 2000 and 4000 respectively.

Reviewer: ```suggestion
In this test, we focus on the latency overhead and resource consumption added by OSM's.  We currently test the per pod RPS at 500, 1000, 2000 and 4000 respectively.
```

Reviewer: let's also add some smaller numbers. People may be interested in 10 and 100

Author: I will add 100. In our testing when the RPS is too low, there were too much noise and their results very close.

### Latency

> Below are the request latency at different RPS levels, comparing the scenarios with and without OSM installed.

Reviewer: ```suggestion
Below is the request latency at different RPS levels with and without OSM installed. Note, with OSM installed the added latency comes from both the client and server sidecar.
```

<p align="center">
  <img src="/docs/images/perf/wo-osm-latency.png" width="650"/>
</p>

<p align="center">
  <img src="/docs/images/perf/w-osm-latency.png" width="650"/>
</p>

> The application without OSM performs quite similar across different RPS levels, although there is an upward trend on the P999 latency. Whereas when OSM is installed, the latency starts to increase drastically when it is above 2000 RPS. Looking at the overhead that OSM adds to the P99 latency, it is less than 2ms for RPS below 2000 and around 14ms and 30ms at 2000 and 4000 RPS respectively.

Reviewer: ```suggestion
The application without OSM performs quite similar across different RPS levels, although there is an upward trend on the 99.9 percentile. Whereas when OSM is installed, the latency starts to increase drastically when it is above 2000 RPS. Looking at the overhead that OSM adds to the P99 latency, it is less than 2ms for RPS below 2000 and around 14ms and 30ms at 2000 and 4000 RPS respectively.
```

Reviewer: we also may want to put some of these numbers in a table.

Reviewer 2: +1 on the table; it's easier to grok the specifics in a table layout. In said table, it may also be useful to have a column for the delta latency between calls without OSM and calls with OSM

Author: I think people won't be interested in the exact numbers in the charts, as the numbers will change for their own applications. The primary purpose is to show the trend. But I do agree the numbers in the charts are crowded.

In this case, we can add a table that compares the p99 latency in scenarios with and w/o OSM.

> The application without OSM performs quite similar across different RPS levels, although there is an upward trend on the P999 latency. Whereas when OSM is installed, the latency starts to increase drastically when it is above 2000 RPS. Looking at the overhead that OSM adds to the P99 latency, it is less than 2ms for RPS below 2000 and around 14ms and 30ms at 2000 and 4000 RPS respectively.

Reviewer: Let's also look at why that is happening... are we seeing sidecar's hitting their max consumption? if so, let's not resource constrain, since that loses fidelity for both resource consumption and latency. 

Let's try bumping up the sidecar resources, or even trying bigger machines to see if that's not the case.

I wouldn't expect envoy to tack on that much latency

> Note that the tesing setup is an ideal case where the application is extremely simple. When the services, mesh topology and mesh configuration become more complicated, the latency is expected to be longer.

Reviewer: ```suggestion
Note that our test setup is very simple, with a single client/server relation. Your mileage may vary with more complex applications, or with different traffic policies set within envoy.
```

## Process

The testing mesh contains two types of services. One is a load generator (modified from [wrk2](https://github.com/giltene/wrk2)), which can send requests at specified rate, load balanced among all other services in the mesh. Currently we test HTTP 1 requests only. The load generator serves as the client side of a request. Throughout our test, there will be at most one load generator service. The other type of service is a simple echo application that replies with whatever it receives from an inbound request. It serves as the server side of a request. We can deploy multiple copies of the echo application in the mesh. In short, the mesh topology is a typical 1-to-N service architecture.

<p align="center">
  <img src="/docs/images/perf/mesh-topo.png" width="650"/>
</p>

## Performance

In this test, we focus on the latency overhead and resource consumption added by the OSM sidecar. We currently test the per pod RPS at 100, 500, 1000, 2000 and 4000 respectively.

### Latency

> Below is the request latency at different RPS levels with OSM installed. The added latency comes from both the client and server sidecars.

Author: There is one for latency without OSM but we need more discussion on it. It is a bit surprising that latency without OSM is higher in some cases. I believe this is because Envoy helps to reuse connections and saves time. Thus it's not a fair to directly comparing latency w/ and w/o OSM, and OSM does not support customizing data plane # of connections now.

> Below is the request latency at different RPS levels with OSM installed. The added latency comes from both the client and server sidecars.

Reviewer 3: Thanks for explaining!

---
title: "Performance and Scalability"
description: "Performance benchmark on control plane and data plane sidecar and scalability test of Open Service Mesh"
type: docs
---

Open Service Mesh aims to create a high-performance service mesh while providing all the essential service mesh features. This document describes the tests performed to evaluate the performance and scalability of OSM.

## Testing environment

This test is performed on a Kubernetes cluster provisioned and managed by [Microsoft Azure Kubernetes Service](https://azure.microsoft.com/en-us/services/kubernetes-service/). The cluster has 100 Standard_DS2_v2 nodes (each node has 2 vCPUs and 7GB memory). It is ensured that there is only one OSM controller pod running with default resource quota during the test. OSM is configured in non-permissive traffic mode.

> ## Process

Reviewer: maybe note that we are using http 1 for this, since http 2 or grpc would have different performance characteristics)

Author: Sure.

Yes we should mentioned this is a minimal setup of the mesh. We will benchmark other popular features of OSM and service mesh in general in the future.

> The testing mesh contains two types of service. One is a load generator (modified from [wrk2](https://github.com/giltene/wrk2)), which can send requests at specified rate, load balanced among all other services in the mesh. The load generator serves as the client side of a request. Throughout our test, there will be at most one load generator service. The other type of service is a simple echo application that replies with whatever it receives from an inbound request. It serves as the server side of a request. We can deploy multiple copies of the echo application in the mesh. In short, the mesh topology is a typical 1-to-N service architecture.

Reviewer: ```suggestion
The testing mesh contains two types of services. One is a load generator (modified from [wrk2](https://github.com/giltene/wrk2)), which can send requests at specified rate, load balanced among all other services in the mesh. The load generator serves as the client side of a request. Throughout our test, there will be at most one load generator service. The other type of service is a simple echo application that replies with whatever it receives from an inbound request. It serves as the server side of a request. We can deploy multiple copies of the echo application in the mesh. In short, the mesh topology is a typical 1-to-N service architecture.
```

<p align="center">
  <img src="/docs/images/perf/mesh-topo.png" width="650"/>
</p>

## Performance

> In this test, we focus on the latency overhead and resource consumption added by OSM's. We currently test the per pod RPS at 100, 500, 1000, 2000 and 4000 respectively.

Reviewer: ```suggestion
In this test, we focus on the latency overhead and resource consumption added by the OSM sidecar. We currently test the per pod RPS at 100, 500, 1000, 2000 and 4000 respectively.
```

### Latency

Below is the request latency at different RPS levels with and without OSM installed. Note, with OSM installed the added latency comes from both the client and server sidecar.

<p align="center">
  <img src="/docs/images/perf/wo-osm-latency.png" width="650"/>
</p>

<p align="center">
  <img src="/docs/images/perf/w-osm-latency.png" width="650"/>
</p>

<table style="width: 100%">
<thead>
<tr>
<th>Per pod RPS</th>
<th>100</th>
<th>500</th>
<th>1000</th>
<th>1500</th>
<th>2000</th>
<th>4000</th>
</tr>
</thead>
<tbody>
<tr>

> <td>P99 latency without OSM</td>

Reviewer: ```suggestion
<td>P99 latency without OSM (ms)</td>
```

<td>2</td>
<td>4.6</td>
<td>3.7</td>
<td>4.23</td>
<td>4.3</td>
<td>3.1</td>
</tr>
<tr>

> <td>P99 latency with OSM</td>

Reviewer: ```suggestion
<td>P99 latency with OSM (ms)</td>
```

> <td>10.1</td>

Reviewer: why is 100 rps so much higher than 500 rps? can we try running this a few times?

Reviewer: are we appropriately chopping off the beginning metrics of the test to account for a stable OSM state?

<td>4.93</td>
<td>5.93</td>
<td>5</td>
<td>17.5</td>

> <td>32.8</td>

Reviewer: You sure these higher numbers at 4000 rps, still aren't due to cpu bottlenecks?

Author: Yes. We tested the 8000 rps and the latency continues to increase.

Author: The sidecar CPU usage is below in the Resource Consumption section that can be compared with the latency.

Reviewer: what limits did you use? I'd also be careful that with the p99.9 latency you may have a spike in CPU usage for brief moment that hits limits, and affects low number of requests to put this over the limit. so we may want to view cpu usage over time for each container

Author: The CPU limit for echo application is 1 core. P99.9 is averaged across the collected metrics in the 10 mins testing time. As our metrics pulling interval is 15s, we will have 60 data points. I don't think it's necessary to show the over-time CPU usage chart in this document, as it's too detailed.

Author: For reference, this is the sidecar CPU usage over a test run. It looks very stable.

![image](https://user-images.githubusercontent.com/872876/160935357-010cdfdb-0158-43bb-8016-6a280099d9a4.png)

Reviewer: can you chart plot the max value? that might be more indicative of anything hitting limits.

what is the container limit btw?

Author: Discussed offline and we noticed that the increase in latency is caused due to the Envoy container CPU limit. Added an explanation in the documentation to call out this observation.

Reviewer: what were the new container limits you set for the new test?

Author: Envoy CPU limit is set to 4 cores.

Reviewer: I'm still a bit skeptical of that 4.6ms on the p999 for latency not being artificial due to one of the sidecars getting throttled. Can we try running again, or even further increasing the number of nodes/cpu limits?

At the limits we set, how many pods did we have? did we have enough CPU in the cluster to satisfy all of the max limits for sidecar/controle plane/load gen/and echo application?

Author: Tests re-run on the cluster and gets more reasonable results. It seems that cluster performance fluctuate days apart. Document is updated.

## Testing environment

This test is performed on a Kubernetes cluster provisioned and managed by [Microsoft Azure Kubernetes Service](https://azure.microsoft.com/en-us/services/kubernetes-service/). The cluster has 100 Standard_DS2_v2 nodes (each node has 2 vCPUs and 7GB memory). It is ensured that there is only one OSM controller pod running with default resource quota during the test. OSM is configured in non-permissive traffic mode.

## Process

The testing mesh contains two types of service. One is a load generator (modified from [wrk2](https://github.com/giltene/wrk2)), which can send requests at specified rate, load balanced among all other services in the mesh. The load generator serves as the client side of a request. Throughout our test, there will be at most one load generator service. The other type of service is a simple echo application that replies with whatever it receives from an inbound request. It serves as the server side of a request. We can deploy multiple copies of the echo application in the mesh. In short, the mesh topology is a typical 1-to-N service architecture.

<p align="center">
  <img src="/docs/images/perf/mesh-topo.png" width="650"/>
</p>

> During the test, the echo applications will be deployed all at the same time. Therefore, the control plane receives a load spike after the deployment starts. All the Envoy sidecars are connecting to the control plane simutaneously. After that, the control plane becomes much more idle. We deliberately create this load spike at test bootstrap time in order to explore the maximum scale that the mesh can handle. Each round of test will last for 15 minutes.

Reviewer 2: This feels like it should be under the Scalability heading. I completely forgot that the services were going to be added in batches until I saw "peak time" on the graphs. It makes more sense to colocate process explanations with the specific tests they're describing

</p>

The application without OSM performs quite similar across different RPS levels, although there is an upward trend on the 99.9 percentile latency. Whereas when OSM is installed, the latency below 1000 RPS does not change much. It starts to increase proportially when latency is above 1000 RPS. Looking at the overhead that OSM adds to the P99 latency, it is less than 2ms for RPS below 2000 and around 14ms and 30ms at 2000 and 4000 RPS respectively.

Note that our test setup is very simple, with a single client/server relation. Your mileage may vary with more complex applications, or with different traffic policies set on the mesh.

### Resource Consumption

Control plane resource consumption is stable as it is not relavant to the data plane request rate.

For the sidecars on the data plane, the memory usage is relatively consistent across different RPS levels, which is around 140MB. The CPU usage has a clear increasing trend and it is nearly proportional to the RPS level. In our case, roughly every 100 RPS takes 0.1 cores of CPU on sidecar container.

<p align="center">
  <img src="/docs/images/perf/avg-cpu.png" width="650"/>
</p>

## Scalability

> In the scalability test, we explore how resource consumption changes for both the control plane and the data plane as the mesh scales up. A number of echo services will be added to the mesh. Each service will have 10 replicas. The mesh is eventually added 200, 400, 800 and 1600 pods respectively in each round of the test. The load generator service is not used in scalability test.

Reviewer 2: ```suggestion
In the scalability test, we explore how resource consumption changes for both the control plane and the data plane as the number workloads in the mesh increases. Over [insert time frame here], an increasing number of echo services will be added to the mesh at a rate of [e.g. 5 services/minute]. Each service will have 10 replicas. After a total of 4 rounds,  measurements will be recorded as the mesh workload count hits 200, 400, 800 and 1600 pods respectively in each round of the test. The load generator service is not used in the scalability test.
```

Author: I will re-word, but the test does not run the way as the suggestion.

We don't specify the service adding rate. The services are deployed using Helm with specified number of services and replicas. I suppose it applies manifest files at the best rate that Kubernetes apiserver can handle.

Reviewer 2: Yeah I realized it was a batch add when I re-read the earlier process paragraph. Please remove whatever is not accurate 👍

### Control Plane

#### CPU Usage

> Below shows the CPU usage of control plane at both pod-churn time and stable time:

Reviewer 3: Do pod-churn, peak, and spike all have the same meaning in this doc?

Author: Good question. Yes they are the same thing. I will make them consistent.
